File: mpitest/funneled/hello.c
#include <stdio.h>
#include <mpi.h>

int main (int argc, char **argv)
{
  int rank, size;
  int name_len;
  char name[MPI_MAX_PROCESSOR_NAME];
  int request = MPI_THREAD_FUNNELED, provided=-1;
  MPI_Init (&argc, &argv);/* starts MPI */
  //MPI_Init_thread (&argc, &argv, request, &provided);
  printf("%d %d\n", request, provided);
#pragma omp parallel
  {
#pragma omp master
    {
      MPI_Comm_rank (MPI_COMM_WORLD, &rank);/* get current process id */
      MPI_Comm_size (MPI_COMM_WORLD, &size);/* get number of processes */
      MPI_Get_processor_name(name, &name_len);
      printf( "Hello, parallel world %d / %d : %s\n", rank, size, name ); fflush(stdout);
      MPI_Barrier(MPI_COMM_WORLD);
    }
    printf("fin\n");
  }
  MPI_Finalize();
  return 0;
}

File: mpitest/funneled/hello.f
      program hello
      include 'mpif.h'
      integer ierr,myproc,hostnm,size
      character*64 hostname
      integer r,p
      r = MPI_THREAD_FUNNELED
      p = -1
      call mpi_init(ierr)
!      call mpi_init_thread(r,p,ierr)
      write(*,*)r,p
!$omp parallel
!$omp master
      call mpi_comm_rank(MPI_COMM_WORLD, myproc, ierr)
      call mpi_comm_size(MPI_COMM_WORLD, size, ierr)
      ierr=hostnm(hostname)
      write(6,100) myproc,size,hostname
      call mpi_barrier(MPI_COMM_WORLD,ierr)
!$omp end master
      write(*,*)"fin"
!$omp end parallel
 100        format(1x,"Hello, parallel world ",i3,"/",i3," : host ",A64)
      call mpi_finalize(ierr)
      end
